#90: Forgetting roots and favorites

ShubertMunthali opened the pull request:
#90

kirillt commented:
If I understand correctly, we need to merge https://github.com/ARK-Builders/ark-filepicker/pull/17 first, and re-build APK for this branch then.

kirillt commented on app/build.gradle:
```diff
@@ -101,7 +101,7 @@ dependencies {
     implementation 'com.davemorrissey.labs:subsampling-scale-image-view-androidx:3.10.0'
     implementation 'com.github.MikeOrtiz:TouchImageView:3.1.1'
-    implementation 'com.github.ARK-Builders:ark-filepicker:ae2c37ea87'
+    implementation 'com.github.ARK-Builders:ark-filepicker:ed2ef717e4'
```
This reference might change after ark-filepicker PR will be merged.

ShubertMunthali replied:
> If I understand correctly, we need to merge [ARK-Builders/ark-filepicker#17](https://github.com/ARK-Builders/ark-filepicker/pull/17) first, and re-build APK for this branch then.

Yes

kirillt commented on app/build.gradle:
```diff
@@ -101,7 +102,7 @@ dependencies {
     implementation 'com.davemorrissey.labs:subsampling-scale-image-view-androidx:3.10.0'
     implementation 'com.github.MikeOrtiz:TouchImageView:3.1.1'
-    implementation 'com.github.ARK-Builders:ark-filepicker:ae2c37ea87'
+    implementation 'com.github.ARK-Builders:ark-filepicker:fde52fe7a5'
```
We should not forget to change this after merge in `ark-filepicker` repo.

ShubertMunthali replied:
Yes yes

kirillt commented on gradle/wrapper/gradle-wrapper.properties:
```diff
@@ -1,6 +1,6 @@
 #Sat Jan 08 19:45:49 CST 2022
 distributionBase=GRADLE_USER_HOME
-distributionUrl=https\://services.gradle.org/distributions/gradle-7.0.2-bin.zip
+distributionUrl=https\://services.gradle.org/distributions/gradle-7.3.3-bin.zip
```
Should this version be the same as in `build.gradle`? I see 7.2.0 there, but maybe not really important.

ShubertMunthali replied:
Where exactly in ```build.gradle```. I have looked, I don't find it. Can you provide a reference please. I think it may be neccessary.

mdrlzy replied:
`distributionUrl` represents `Gradle` version
`com.android.tools.build:gradle` represents `Android Gradle plugin` version
Gradle is a bit chaotic

kirillt replied:
@mdrlzy so all is good here?

mdrlzy replied:
yep

kirillt commented:
Not sure why is it so
```
[kirill@lenovo tmp]$ unzip release-arm64-v8a-apk.zip 
Archive:  release-arm64-v8a-apk.zip
  End-of-central-directory signature not found.  Either this file is not
  a zipfile, or it constitutes one disk of a multi-part archive.  In the
  latter case the central directory and zipfile comment will be found on
  the last disk(s) of this archive.
unzip:  cannot find zipfile directory in one of release-arm64-v8a-apk.zip or
        release-arm64-v8a-apk.zip.zip, and cannot find release-arm64-v8a-apk.zip.ZIP, period.
```

ShubertMunthali commented:
> Not sure why is it so
> 
> ```
> [kirill@lenovo tmp]$ unzip release-arm64-v8a-apk.zip 
> Archive:  release-arm64-v8a-apk.zip
>   End-of-central-directory signature not found.  Either this file is not
>   a zipfile, or it constitutes one disk of a multi-part archive.  In the
>   latter case the central directory and zipfile comment will be found on
>   the last disk(s) of this archive.
> unzip:  cannot find zipfile directory in one of release-arm64-v8a-apk.zip or
>         release-arm64-v8a-apk.zip.zip, and cannot find release-arm64-v8a-apk.zip.ZIP, period.
> ```

What happened, can you provide more scope, I should look for the cause.

kirillt commented:
@ShubertMunthali https://github.com/ARK-Builders/ark-filepicker/pull/17 is merged.

Let's change Git references in this PR, rebuild the APK and I'll test it.
We'll see if Zip error is gone.

ShubertMunthali commented:
> @ShubertMunthali https://github.com/ARK-Builders/ark-filepicker/pull/17 is merged.
> 
> Let's change Git references in this PR, rebuild the APK and I'll test it.
> We'll see if Zip error is gone.

Alright

kirillt commented:
- [ ] Roots disappear only after restarting the app
  * delete a root
  * nothing happens
  * restart Navigator
  * the root is gone
- [ ] Re-introduced roots cause the app crashing
  * delete a root, e.g. `/a/b/c`
  * `/a/b/c` disappears both from Navigator and any external file browser
  * externally create new folder by the same path `/a/b/c`
  * Navigator crashes, it can't send the report as well
  * externally remove `/a/b/c`
  * Navigator starts with 1 crash which can be ignored

kirillt commented:
Forgetting and deletion of favorites looks good.

kirillt commented:
Works as expected, but I encountered this exception after I deleted a root, switched to another app and then restored the app from background:
```
java.nio.file.NoSuchFileException: /storage/emulated/0/TEST-1/.ark/stats/tag-labeled-n
	at sun.nio.fs.UnixFileSystemProvider.newByteChannel(UnixFileSystemProvider.java:214)
	at java.nio.file.spi.FileSystemProvider.newOutputStream(FileSystemProvider.java:434)
	at java.nio.file.Files.newOutputStream(Files.java:216)
	at kotlin.io.path.PathsKt__PathReadWriteKt.writeText(PathReadWrite.kt:171)
	at kotlin.io.path.PathsKt__PathReadWriteKt.writeText$default(PathReadWrite.kt:170)
	at space.taran.arknavigator.mvp.model.repo.stats.category.TagLabeledNStorage.flush(TagLabeledNStorage.kt:67)
	at space.taran.arknavigator.mvp.model.repo.stats.category.StatsCategoryStorage$flushFlow$1$1.invokeSuspend(StatsCategoryStorage.kt:23)
	at space.taran.arknavigator.mvp.model.repo.stats.category.StatsCategoryStorage$flushFlow$1$1.invoke(Unknown Source:8)
	at space.taran.arknavigator.mvp.model.repo.stats.category.StatsCategoryStorage$flushFlow$1$1.invoke(Unknown Source:4)
	at kotlinx.coroutines.flow.FlowKt__TransformKt$onEach$$inlined$unsafeTransform$1$2.emit(Emitters.kt:223)
	at kotlinx.coroutines.flow.FlowKt__DelayKt$debounceInternal$1$3$1.invokeSuspend(Delay.kt:233)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
	at kotlinx.coroutines.internal.LimitedDispatcher.run(LimitedDispatcher.kt:42)
	at kotlinx.coroutines.scheduling.TaskImpl.run(Tasks.kt:95)
	at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:570)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:749)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:677)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:664)
	Suppressed: kotlinx.coroutines.DiagnosticCoroutineContextException: [StandaloneCoroutine{Cancelling}@8526327, Dispatchers.IO]
```

ShubertMunthali commented:
> Works as expected, but I encountered this exception after I deleted a root, switched to another app and then restored the app from background:
> ```
> java.nio.file.NoSuchFileException: /storage/emulated/0/TEST-1/.ark/stats/tag-labeled-n
> 	at sun.nio.fs.UnixFileSystemProvider.newByteChannel(UnixFileSystemProvider.java:214)
> 	at java.nio.file.spi.FileSystemProvider.newOutputStream(FileSystemProvider.java:434)
> 	at java.nio.file.Files.newOutputStream(Files.java:216)
> 	at kotlin.io.path.PathsKt__PathReadWriteKt.writeText(PathReadWrite.kt:171)
> 	at kotlin.io.path.PathsKt__PathReadWriteKt.writeText$default(PathReadWrite.kt:170)
> 	at space.taran.arknavigator.mvp.model.repo.stats.category.TagLabeledNStorage.flush(TagLabeledNStorage.kt:67)
> 	at space.taran.arknavigator.mvp.model.repo.stats.category.StatsCategoryStorage$flushFlow$1$1.invokeSuspend(StatsCategoryStorage.kt:23)
> 	at space.taran.arknavigator.mvp.model.repo.stats.category.StatsCategoryStorage$flushFlow$1$1.invoke(Unknown Source:8)
> 	at space.taran.arknavigator.mvp.model.repo.stats.category.StatsCategoryStorage$flushFlow$1$1.invoke(Unknown Source:4)
> 	at kotlinx.coroutines.flow.FlowKt__TransformKt$onEach$$inlined$unsafeTransform$1$2.emit(Emitters.kt:223)
> 	at kotlinx.coroutines.flow.FlowKt__DelayKt$debounceInternal$1$3$1.invokeSuspend(Delay.kt:233)
> 	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
> 	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
> 	at kotlinx.coroutines.internal.LimitedDispatcher.run(LimitedDispatcher.kt:42)
> 	at kotlinx.coroutines.scheduling.TaskImpl.run(Tasks.kt:95)
> 	at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:570)
> 	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:749)
> 	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:677)
> 	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:664)
> 	Suppressed: kotlinx.coroutines.DiagnosticCoroutineContextException: [StandaloneCoroutine{Cancelling}@8526327, Dispatchers.IO]
> ```

Seems this happened in stats. I think stats doesn't know the root is gone

mdrlzy commented:
> Works as expected, but I encountered this exception after I deleted a root, switched to another app and then restored the app from background:
> 
> ```
> java.nio.file.NoSuchFileException: /storage/emulated/0/TEST-1/.ark/stats/tag-labeled-n
> 	at sun.nio.fs.UnixFileSystemProvider.newByteChannel(UnixFileSystemProvider.java:214)
> 	at java.nio.file.spi.FileSystemProvider.newOutputStream(FileSystemProvider.java:434)
> 	at java.nio.file.Files.newOutputStream(Files.java:216)
> 	at kotlin.io.path.PathsKt__PathReadWriteKt.writeText(PathReadWrite.kt:171)
> 	at kotlin.io.path.PathsKt__PathReadWriteKt.writeText$default(PathReadWrite.kt:170)
> 	at space.taran.arknavigator.mvp.model.repo.stats.category.TagLabeledNStorage.flush(TagLabeledNStorage.kt:67)
> 	at space.taran.arknavigator.mvp.model.repo.stats.category.StatsCategoryStorage$flushFlow$1$1.invokeSuspend(StatsCategoryStorage.kt:23)
> 	at space.taran.arknavigator.mvp.model.repo.stats.category.StatsCategoryStorage$flushFlow$1$1.invoke(Unknown Source:8)
> 	at space.taran.arknavigator.mvp.model.repo.stats.category.StatsCategoryStorage$flushFlow$1$1.invoke(Unknown Source:4)
> 	at kotlinx.coroutines.flow.FlowKt__TransformKt$onEach$$inlined$unsafeTransform$1$2.emit(Emitters.kt:223)
> 	at kotlinx.coroutines.flow.FlowKt__DelayKt$debounceInternal$1$3$1.invokeSuspend(Delay.kt:233)
> 	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
> 	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
> 	at kotlinx.coroutines.internal.LimitedDispatcher.run(LimitedDispatcher.kt:42)
> 	at kotlinx.coroutines.scheduling.TaskImpl.run(Tasks.kt:95)
> 	at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:570)
> 	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:749)
> 	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:677)
> 	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:664)
> 	Suppressed: kotlinx.coroutines.DiagnosticCoroutineContextException: [StandaloneCoroutine{Cancelling}@8526327, Dispatchers.IO]
> ```

The problem is that `StatsStorage` does not write to the disk immediately, but only if there were no events for 10 seconds.
In theory, after removing root, we should clear links to all storages, but now it is problematic to do this, since we are moving them to arklib
Let's do try catch for now)
